### DISCUSSION TITLE
Question: export_png requires dependencies that 'save' button does not

### DISCUSSION THREAD

**Discussion by Bob-May** [bokeh/bokeh#11576] [General]
Why does export_png() require installation of selenium geckodriver firefox dependencies

When the 'save' button on a figure's toolbar can save a png without these dependencies?

**Comment by bryevdv** [comment]
@Bob-May Please use GH issues only for bug reports or feature requests, not general questions or discussions. 

First of all: Bokeh is a ***browser plotting library***, full stop. it happens to have a Python API to facilitate creating rendering in the browser, but all Bokeh output is always and only ever rendered in a browser. Nothing is ever rendered in Python. 

With that in mind:

Clicking the `SaveTool` toolbar button excutes *JavaScript code,  in the browser*, and is therefore able to access the HTML canvas (i.e. where the rendering actually happens) directly. 

By contrast `export_png` is a Python API. There is nothing rendered anywhere in Python (see above) so the only option is to send to Bokeh document to a headless browser for rendering, and then to capture the result of that. Currently, Selenium is used to make that happen, tho in the future there may be an option to use chromium dev protocol, which would eliminate the need for Selenium.

**Reply by Bob-May** [reply]
@bryevdv 
Thanks for the explaination.

So, when using the Bokeh Python API in a Jupyter Notebook, is there a way to invoke the SaveTool via Python?

**Reply by bryevdv** [reply]
There is nothing built in to do that (and won't be). In general, it's not impossible to inject and execute JS code into the notebook page from Python code in notebook cells [1], so it's also not impossible to imagine finding some way to do what you suggest. But it would be fragile, might break across different notebook versions, or between classic notebook and jupyter-lab, or require different handling on different browsers, or even just not be able to work at all on some browsers etc.

[1] It is in fact impossible on some jupyter-lab versions, at least without extra extensions manually installed. These kinds of uncertainties are why we would never pursue this kind of approach.
